### ISSUE TITLE
weird interactions in gradient boosting and successive halving

### ISSUE THREAD

**Issue by amueller** [dabl/dabl#84]
```python
from sklearn.experimental import enable_hist_gradient_boosting
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.datasets import fetch_openml
from dabl.search import RandomSuccessiveHalving
from scipy import stats

bunch = fetch_openml(data_id=40701)

X = bunch.data
y = bunch.target
param_dist = {'max_leaf_nodes': stats.randint(2, 100)}
rsh = RandomSuccessiveHalving(HistGradientBoostingClassifier(),
                              param_distributions=param_dist,
                              verbose=10, r_min=40, scoring="roc_auc")
rsh.fit(X, y)
```
doesn't work as expected. In the first round all the AUCs are .5 and I think some random params get selected.
I'm not sure what's happening. It could be that because it's a imbalanced dataset something weird happens? Also without specifying ``r_min`` it doesn't work at all b/c of validation set issues, I think.

@NicolasHug could you have a look?
Also: I realized HistGradientBoosting doesn't have all the subsampling options and no support for class_weight ;)

**Comment by NicolasHug** [comment]
>In the first round all the AUCs are .5 

That's because the training set is too small at this stage and min_samples_leaf=20 by default. All the trees only have one node.
We've had discussions about whether this is a sensible default with @ogrisel.

>Also without specifying r_min it doesn't work at all b/c of validation set issues

That's not related to the GBDTs, it happens with other estimators. So that's probably our magic `r_min` that isn't as magic as we'd hope. I'll look into it.

**Comment by ogrisel** [comment]
With could have a `min_samples_leaf="auto"` that would be set to `min(X.shape[0], 20)` at fit time.

Question @NicolasHug does LightGBM behave the same on small datasets?

**Comment by ogrisel** [comment]
+1 for implementing for upping the priority on `sample_weights` :)

For imbalanced classes it would be useful to implement undersampling of the majority class for each boosting round, but maybe by subclassing in imbalanced-learn instead of scikit-learn as I am not sure if it's a standard practice or not. That would require heavy benchmarking to show whether this is a empirically useful strategy or not  (/cc @glemaitre).

**Comment by NicolasHug** [comment]
> Question @NicolasHug does LightGBM behave the same on small datasets?

yes their default is 20 (tried with an lgtm estimator, AUC is also .5)

**Comment by NicolasHug** [comment]
About `r_min='auto'`, we set it to `n_classes * n_splits * 2`

we end up with splits where only one class is present since the classes are imbalanced.

Maybe a more sensible default would be something like 
``n_classes * n_splits / prop_least_frequent_class` ?

**Comment by amueller** [comment]
@NicolasHug yes that makes more sense, I think?
@ogrisel balanced subsampling per tree in gbrt is an obvious strategy that seems promising. Do you know if anyone has published it?

I actually wasn't able to find a good parameter setting for this dataset. Is there one that works on small subsamples of the data?

I think there are several bad things happening here, maybe we should open separate issues ;) I think the ``r_min`` issue is pretty clear.

I think we should also guard against too many ties in the first iteration of SH because that basically creates random results. Maybe that's fine sometimes? We should at least warn, I think? If I hadn't dug in, I wouldn't have seen that some parts of the parameter space are arbitrarily discarded. However, if you started with a random sample of the space, basically you're just skipping the first iteration and doubling r_min.

WRT gbrt defaults, I think we should try to be as robust as possible. But will gbrt do something sensible with one or two samples per class? I guess if you have two classes of 10 samples each you might be able to learn something in principle.

**Comment by ogrisel** [comment]
@ogrisel balanced subsampling per tree in gbrt is an obvious strategy that seems promising. Do you know if anyone has published it?

No.
